tests/xdp/xdp_tests.cpp

```diff
@@ -117,31 +52,219 @@ TEST_CASE("xdp_reflect_test", "[xdp_tests]")
+        error = WSAWaitForMultipleEvents(1, &overlapped.hEvent, TRUE, 1000, TRUE);
+        if (error != WSA_WAIT_EVENT_0)
+            if (error == WSA_WAIT_TIMEOUT)
+                FAIL("Receiver socket did not receive any message in 1 second.");
+            else
+                FAIL("Waiting on receiver socket failed with " << error);
+
+        if (!WSAGetOverlappedResult(
+                socket,
+                &overlapped,
+                reinterpret_cast<LPDWORD>(&bytes_received),
+                FALSE,
+                reinterpret_cast<LPDWORD>(&recv_flags)))
+            FAIL("WSArecvFrom on the receiver socket failed with error: " << WSAGetLastError());
+    }
+
+    void
+    get_received_message(uint32_t& message_size, _Out_ char*& message)
+    {
+        message_size = bytes_received;
+        message = recv_buffer.data();
+    }
+
+    void
+    get_sender_address(PSOCKADDR& from, int& from_length)
+    {
+        from = (PSOCKADDR)&sender_address;
+        from_length = sender_address_size;
+    }
+
+  private:
+    SOCKET socket;
+    int sock_type;
+    int protocol;
+    WSAOVERLAPPED overlapped;
+    std::vector<char> recv_buffer;
+    uint32_t recv_flags;
+    sockaddr_storage sender_address;
+    int sender_address_size;
+    uint32_t bytes_received = 0;
+} receiver_socket_t;
+
+TEST_CASE("xdp_encap_reflect_test", "[xdp_tests]")
+{
+    // Initialize the remote address.
+    struct sockaddr_storage remote_address = {};
+    ADDRESS_FAMILY address_family;
+    _get_remote_address(remote_address, &address_family);
+    REQUIRE((address_family == AF_INET || address_family == AF_INET6));
+    int protocol = (address_family == AF_INET) ? IPPROTO_IPV4 : IPPROTO_IPV6;
+    // Create a RAW receiver socket with protocol being IPV4 or IPV6 based on the address family of the remote host.
+    receiver_socket_t receiver_socket(SOCK_RAW, protocol);
+    // Post an asynchronous receive on the receiver socket.
+    receiver_socket.post_async_receive();
+    // send message to remote host on reflection port.
+    const char* message = "Bo!ng";
+    _send_message_to_remote_host(message, remote_address);
+    // Complete the asynchronous receive and obtain the reflected message.
+    receiver_socket.complete_async_receive();
+    // Verify if the received message was reflected by remote host.
+    PSOCKADDR sender_address = nullptr;
+    int sender_length = 0;
+    receiver_socket.get_sender_address(sender_address, sender_length);
+    REQUIRE(INET_ADDR_EQUAL(
+        remote_address.ss_family, INETADDR_ADDRESS((PSOCKADDR)&remote_address), INETADDR_ADDRESS(sender_address)));
+    // Verify if the received message is expected.
+    uint32_t bytes_received = 0;
+    char* received_message = nullptr;
+    receiver_socket.get_received_message(bytes_received, received_message);
+    if (address_family == AF_INET) {
+        // Raw sockets with protocol IPPROTO_IPV4 receives outer IP header in received message.
```

Comment:
```suggestion
        // Raw socket with protocol IPPROTO_IPV4 received outer IP header in received message.
```

Reply:
I meant to describe the behavior of raw sockets in general - and not what this instance of socket is doing.